modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/delegate/delegate.es.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+/**
+ * Checks if element is disabled or whether it exists in a disabled element tree
+ * @param {!Element} element The DOM element to check.
+ * @return {boolean}
+ */
+function isDisabled(node) {
+	return !!(node.disabled || node.closest('[disabled]'));
+}
+
+/**
+ * Listens to the specified event on the given DOM element, but only calls the
+ * given callback listener when it's triggered by elements that match the
+ * given selector or target element.
+ * @param {!Element} element The DOM element the event should be listened on.
+ * @param {string} eventName The name of the event to listen to.
+ * @param {!Element|string} selectorOrTarget Either an element or css selector
+ *     that should match the event for the listener to be triggered.
+ * @param {!function(!Object)} callback Function to be called when the event
+ *     is triggered. It will receive the normalized event object.
+ * @return {!EventHandle} Can be used to remove the listener.
+ */
+function delegate(element, eventName, selectorOrTarget, callback) {
+	const eventHandler = (event) => {
+		if (eventName === 'click' && isDisabled(event.target)) {
+			return;
+		}
+
+		if (
+			event.target === selectorOrTarget ||
+			(typeof selectorOrTarget === 'string' &&
+				event.target.matches(selectorOrTarget))
+		) {
+			callback(event);
+		}
+	};
+
+	element.addEventListener(eventName, eventHandler);
+
+	return () => {
```

Comment:
I'm torn about this... on one hand, I see this being nice if used as

```javascript
useEffect(() => {
    return delegate(...);
});
```

But I'm not entirely sure about it when used with something else compared with the traditional `handle.dispose` or `handle.detach` approach.

```javascript
useEffect(() => {
    const removeDelegate = delegate(...);

    const removeDelegates = [
        delegate(...),
        delegate(...)
    ];

    const handler = () => {};
    window.addEventListener('click', handler);

    return () => {
        // weird function call
        removeDelegate();

        // weird array stack to dispose of them?
        removeDelegates.forEach(removeDelegate => removeDelegate());

        window.removeEventListener('click', handler);
    }
});

// Real code in `frontend-js-tooltip-support-web`
useEffect(() => {
	const TRIGGER_SHOW_HANDLES = TRIGGER_SHOW_EVENTS.map((eventName) => {
		return delegate(...);
	});

	const TRIGGER_HIDE_HANDLES = TRIGGER_HIDE_EVENTS.map((eventName) => {
		return delegate(...);
	});

	const TOOLTIP_ENTER = delegate(...);
	const TOOLTIP_LEAVE = delegate(...);

	return () => {
		[
			TOOLTIP_ENTER,
			TOOLTIP_LEAVE,
			...TRIGGER_HIDE_HANDLES,
			...TRIGGER_SHOW_HANDLES,
		].forEach((removeDelegate) => removeDelegate());
                // before, `handle.dispose` seemed a bit more obvious 🤷 
                ].forEach((handle) => handle.dispose());
	};
}, [state]);
```

Thoughs?

Reply:
I'll investigate today and see if I can some up with something that fits an existing patten in dxp

Reply:
I'm not super opinionated about this, just want to make sure this is ergonomic when used in some of the existing places we have it now like shown above or in combination with `window.addEventListener`  that would need to also be cleaned up.